src/ImmutableProperties.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Codelicia\Immutable;
+
+use ReflectionObject;
+use ReflectionProperty;
+use ReflectionType;
+use RuntimeException;
+use function array_key_exists;
+use function class_implements;
+use function get_class;
+use function gettype;
+use function in_array;
+use function interface_exists;
+use function is_object;
+use function spl_object_hash;
+use function sprintf;
+
+trait ImmutableProperties
+{
+    /** @return string[][]|ReflectionProperty[][] */
+    protected function &reflectionProperty(): array
+    {
+        static $reflectionProperty = [];
+
+        if (! array_key_exists(spl_object_hash($this), $reflectionProperty)) {
+            $reflectionProperty[spl_object_hash($this)] = [];
+        }
+
+        return $reflectionProperty[spl_object_hash($this)];
+    }
+
+    public function __construct()
+    {
+        $this->init();
+    }
+
+    public function init(): void
```

Comment:
I think it should be set as `protected` or even `private`. What if there are multiple calls to this method?